### PR TITLE
feat: Add LSP command to profile opcodes in vscode

### PR DESCRIPTION
**Pull request by kobyhallx**
# Description

## Problem\*

Resolves 

feat: Expose Profile information in vscode #3501

## Summary\*

Prepares lSP infrastructure to handle custom profile command. This command is responsible for unconditionally compiling Noir workspace and resulting in opcode counts per source code span which will later be accumulated into lines to be presented.

This is improvement over accessing this information through CLI tool.

## Documentation\*

Check one:
- [x] No documentation needed.
- [ ] Documentation included in this PR.
- [ ] **[Exceptional Case]** Documentation to be submitted in a separate PR.

# PR Checklist\*

- [x] I have tested the changes locally.
- [x] I have formatted the changes with [Prettier](https://prettier.io/) and/or `cargo fmt` on default settings.

### CODE REVIEW

File: tooling/lsp/src/requests/profile_run.rs [hunk 1]
```diff
@@ -0,0 +1,104 @@
...
+        Some(_package) => {
+            // let (mut _context, crate_id) = prepare_package(package, Box::new(get_non_stdlib_asset));
+
+            let (binary_packages, contract_packages): (Vec<_>, Vec<_>) = workspace
+                .into_iter()
+                .filter(|package| !package.is_library())
+                .cloned()
+                .partition(|package| package.is_binary());
+
+            let is_opcode_supported = |_opcode: &Opcode| true;
```
Comment (guipublic) [inline tooling/lsp/src/requests/profile_run.rs:63]: can we get the correct backend instead?

File: tooling/nargo/src/ops/compile.rs [hunk 2]
```diff
@@ -0,0 +1,154 @@
...
+                compile_options.deny_warnings,
+                compile_options.silence_warnings,
+            )
+        })
+        .collect::<Result<_, _>>()?;
+
+    Ok((compiled_programs, compiled_contracts))
+}
+
+pub fn compile_program(
```
Comment (guipublic) [inline tooling/nargo/src/ops/compile.rs:66]: is this code duplicated with cargo_cli?
if yes, can we have it only in one place?

**Review by guipublic — COMMENTED** [review]
LGTM, I have a few questions though.

File: tooling/nargo/src/ops/compile.rs (hunk 2 above)
Comment (kobyhallx) [reply tooling/nargo/src/ops/compile.rs:66]: yes and no, cargo_cli also considers caching while this does not. It looks like we have more refactoring on the way and for that reason I decided to go with "as is" for time being to deliver something working and address debt soonish, would offer #3503 instead?

File: tooling/lsp/src/requests/profile_run.rs [hunk 3]
```diff
@@ -0,0 +1,104 @@
...
+    )
+    .map_err(|err| {
+        // If we found a manifest, but the workspace is invalid, we raise an error about it
+        ResponseError::new(ErrorCode::REQUEST_FAILED, err)
+    })?;
+
+    // Since we filtered on crate name, this should be the only item in the iterator
+    match workspace.into_iter().next() {
+        Some(_package) => {
+            // let (mut _context, crate_id) = prepare_package(package, Box::new(get_non_stdlib_asset));
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply tooling/lsp/src/requests/profile_run.rs]: addressed ad0a466: chore: drop leftover comments

File: tooling/lsp/src/requests/profile_run.rs [hunk 4]
```diff
@@ -0,0 +1,104 @@
...
+        ResponseError::new(ErrorCode::REQUEST_FAILED, "Could not find project root")
+    })?;
+
+    let toml_path = find_package_manifest(root_path, root_path).map_err(|err| {
+        // If we cannot find a manifest, we can't run the test
+        ResponseError::new(ErrorCode::REQUEST_FAILED, err)
+    })?;
+
+    let crate_name = params.package;
+    // let function_name = params.id.function_name();
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply tooling/lsp/src/requests/profile_run.rs]: ad0a466: chore: drop leftover comments

File: tooling/lsp/src/requests/profile_run.rs (hunk 1 above)
Comment (guipublic) [reply tooling/lsp/src/requests/profile_run.rs:60]: It's not super important but it is still relevant. May be add a todo explaining we should use the correct backend eventually?

File: tooling/nargo/src/ops/compile.rs [hunk 5]
```diff
@@ -0,0 +1,154 @@
...
+use crate::prepare_package;
+use crate::{package::Package, workspace::Workspace};
+
+use rayon::prelude::*;
+
+/// Compiles workspace.
+///
+/// # Errors
+///
+/// This function will return an error if .
```
[the remark this answers is not in the document]
Comment (kobyhallx) [reply tooling/nargo/src/ops/compile.rs]: d73167e: chore: address missing error explanation

**Review by guipublic — APPROVED** [review]

**Pull request merged**